### PR TITLE
Rspace/new hot store

### PR DESCRIPTION
**Pull request by lukasz-golebiewski**
## Overview
Part of the RSpace rewrite epic. Aims to provide a new in-memory HotStore lazily reading data from history when needed

### JIRA ticket:
https://rchain.atlassian.net/browse/RCHAIN-3141

### Notes
<sup>_Optional. Add any notes on caveats, approaches you tried that didn't work, or anything else._</sup>

### Please make sure that this PR:
- [x] is at most 200 lines of code (excluding tests),
- [x] meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards),
- [x] includes tests for all added features,
- [x] has a reviewer assigned,
- [x] has [all commits signed](https://rchain.atlassian.net/wiki/spaces/DOC/pages/498630673/How+to+sign+commits+to+rchain+rchain).

### [Bors](https://bors.tech/) cheat-sheet:

- `bors r+` runs integration tests and merges the PR (if it's approved),
- `bors try` runs integration tests for the PR,
- `bors delegate+` enables non-maintainer PR authors to run the above.

### CODE REVIEW

File: rspace/src/main/scala/coop/rchain/rspace/HotStore.scala [hunk 1]
```diff
@@ -59,6 +62,17 @@ private class InMemHotStore[F[_]: Monad, C, P, A, K](
...
     } yield ()
+
+  def getJoins(channel: C): F[List[List[C]]] =
+    for {
+      cache <- S.get
+      res <- cache.joins.get(channel) match {
+              case None        => HR.getJoins(channel)
+              case Some(joins) => Applicative[F].pure(joins)
+            }
+      updatedCache = cache.copy(joins = cache.joins + (channel -> res))
```
Comment (dzajkowski) [inline rspace/src/main/scala/coop/rchain/rspace/HotStore.scala]: this is a needless copy if cache contains the value already, right?

File: rspace/src/main/scala/coop/rchain/rspace/HotStore.scala [hunk 2]
```diff
@@ -40,12 +42,29 @@ private class InMemHotStore[F[_]: Monad, C, P, A, K](
...
+
+  def getData(channel: C): F[List[Datum[A]]] =
+    for {
+      cache <- S.get
+      res <- cache.data.get(channel) match {
+              case None       => HR.getData(channel)
+              case Some(data) => Applicative[F].pure(data)
+            }
+      updatedCache = cache.copy(data = cache.data + (channel -> res))
+      _            <- S.set(updatedCache)
```
Comment (dzajkowski) [inline rspace/src/main/scala/coop/rchain/rspace/HotStore.scala]: this 'get' 'set' mechanism won't work if the monad state is shared. and RSpace is shared between many rholang threads. (hence the double lock)

Comment (lukasz-golebiewski) [reply rspace/src/main/scala/coop/rchain/rspace/HotStore.scala]: Good catch. Due to the double lock it is safe to assume that no two threads are going to modify the data under the same channels. I'm going to add a `TrieMap` holding a cache instance per a list of channels

Comment (lukasz-golebiewski) [reply rspace/src/main/scala/coop/rchain/rspace/HotStore.scala]: BTW, it depends on the underlying `MonadState` implementation. If, as it is currently done in tests, it is an `MVar` it is going to work

**Comment by lukasz-golebiewski** [comment]
Closing in favor of https://github.com/rchain/rchain/pull/2335. The experiments conducted here might come in handy after RSpace is rewritten and benchmark's results are available

**Pull request closed without merging**